### PR TITLE
[tune] Respect max_concurrent_trials for custom searchers

### LINKED ISSUE #5611: Optuna integration significantly reduces performance and limits number of parallel trials in Ray Tune
### Expected behavior

Optuna integration should maintain performance comparable to other search algorithms, utilize CPU resources efficiently, and allow for multiple trials as specified.

### Environment

Ray version: 2.32.0
Python version: 3.11.8
Operating System: Linux-6.6.36.3-microsoft-standard-WSL2-x86_64-with-glibc2.35
Optuna version: 3.6.1

Detailed OS Information:
OS: Linux
OS Version: https://github.com/ray-project/ray/pull/1 SMP PREEMPT_DYNAMIC Sat Jun 29 07:01:04 UTC 2024
OS Release: 6.6.36.3-microsoft-standard-WSL2
Machine: x86_64
Processor: x86_64

### Error messages, stack traces, or logs

```shell
No specific error.
```

### Steps to reproduce

The execution time for 100 trials shows a dramatic difference:

With Optuna: 41 seconds
Without Optuna: 9 seconds

```python
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import ray
from ray import tune
from ray.tune.search.optuna import OptunaSearch
from ray import train, tune

def generate_data(n_samples=1000):
    X = np.random.rand(n_samples, 5)
    y = 2 * X[:, 0] + 3 * X[:, 1] - X[:, 2] + 0.5 * X[:, 3] - 1.5 * X[:, 4] + np.random.normal(0, 0.1, n_samples)
    return X, y

def train_random_forest(config):
    X, y = generate_data()
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2)
    
    rf = RandomForestRegressor(
        n_estimators=config["n_estimators"],
    )
    
    rf.fit(X_train, y_train)
    y_pred = rf.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    
    train.report({"mean_squared_error": mse,})


def main():
    ray.init()
    
    config = {
        "n_estimators": tune.randint(10, 200),
    }
        
    analysis = tune.run(
        train_random_forest,
        config=config,
        num_samples=100, 
        search_alg=OptunaSearch(),
        metric="mean_squared_error",
        mode="min",
        reuse_actors=True
    )
    
    print("Best config:", analysis.best_config)
    print("Best MSE:", analysis.best_result["mean_squared_error"])

if __name__ == "__main__":
    main()

```

### Additional context (optional)

Also referenced here in ray

[](https://github.com/ray-project/ray/issues/46965)

### PR DESCRIPTION
**Pull request by saschwartz**
## Why are these changes needed?

Custom search algorithms (OptunaSearch, HyperOpt, Ax, etc.) run trials **sequentially** even when `max_concurrent_trials` is set, because `_get_max_pending_trials()` hardcodes `max_pending_trials=1` for all non-`BasicVariantGenerator` searchers.

When a user sets `max_concurrent_trials=N`, the searcher is wrapped in `ConcurrencyLimiter(searcher, max_concurrent=N)`, but the pending trial limit ignores this and still returns 1. This means the scheduler only ever queues one trial at a time, regardless of available cluster resources.

The only workaround is the undocumented env var `TUNE_MAX_PENDING_TRIALS_PG`, which most users don't know about.

### Verified impact (Ray 2.54.1, 4 workers, 16 trials × 2s sleep, TPE sampler):

| Config | Time | Result |
|--------|------|--------|
| `max_concurrent_trials=8` (before fix) | 60.4s | sequential |
| `max_concurrent_trials=8` (after fix) | 12.0s | **parallel** |

## Changes

In `_get_max_pending_trials()`: when the search algorithm wraps a `ConcurrencyLimiter` (via `max_concurrent_trials`), extract and use its `max_concurrent` value as the pending trial limit. Without `max_concurrent_trials`, behavior is unchanged (`return 1`). The `TUNE_MAX_PENDING_TRIALS_PG` env var override continues to work.

**Total change: 3 lines of logic + 1 import.**

## Related issue numbers

Fixes https://github.com/optuna/optuna/issues/5611
Related: https://github.com/optuna/optuna/issues/5958
Related: https://github.com/ray-project/ray/pull/48547

## Checks

- [x] I've signed all my commits with `--signoff` (DCO)
- [x] I've run `pre-commit run ruff -a` — all hooks pass
- [x] I've added tests in `test_controller_search_alg_integration.py`
- [x] Changes are backward compatible (no behavior change without `max_concurrent_trials`)

### CODE REVIEW

File: python/ray/tune/execution/tune_controller.py [hunk 1]
```diff
@@ -2116,10 +2120,13 @@ def _get_max_pending_trials(search_alg: SearchAlgorithm) -> int:
...
-    # This is because we don't want to generate too many trials
-    # before we fit the searcher model.
+    # For custom searchers, respect max_concurrent_trials if the user set it.
+    # When max_concurrent_trials is specified, the searcher is wrapped in
+    # SearchGenerator(ConcurrencyLimiter(searcher, max_concurrent=N)).
     if not isinstance(search_alg, BasicVariantGenerator):
+        searcher = getattr(search_alg, "searcher", None)
+        if isinstance(searcher, ConcurrencyLimiter):
+            return searcher.max_concurrent
         return 1
```
[the remark this answers is not in the document]
Comment (saschwartz) [reply python/ray/tune/execution/tune_controller.py:2132]: Applied — now traverses the wrapper chain in a loop. See 2ed0daa.

File: python/ray/tune/tests/execution/test_controller_search_alg_integration.py [hunk 2]
```diff
@@ -380,5 +380,46 @@ def num_running_trials():
...
+        assert _get_max_pending_trials(sg) == 42
+
+    def test_search_generator_without_concurrency_limiter(self):
+        sg = SearchGenerator(_MockSearcher())
+        assert _get_max_pending_trials(sg) == 1
+
+    def test_search_generator_with_concurrency_limiter(self):
+        limited = ConcurrencyLimiter(_MockSearcher(), max_concurrent=8)
+        sg = SearchGenerator(limited)
+        assert _get_max_pending_trials(sg) == 8
```
[the remark this answers is not in the document]
Comment (saschwartz) [reply python/ray/tune/tests/execution/test_controller_search_alg_integration.py:407]: Added `test_search_generator_with_nested_concurrency_limiter` covering `Repeater(ConcurrencyLimiter(...))`. See 2ed0daa.

**Review by pseudo-rnd-thoughts — APPROVED** [review]
Thanks for the PR @saschwartz, this looks good to me. Was there any documentation for ConcurrentLimiter only being applied to BasicVariantGenerator? 
Don't worry about the CI failing for now, we're working on a fix for that

**Comment by saschwartz** [comment]
Hi @pseudo-rnd-thoughts , thanks for the review. 

As to your question: I don't think there is documentation that `max_concurrent_trials` is ignored for the pending trial queue with custom searchers.

Separately, I noticed that The env var `TUNE_MAX_PENDING_TRIALS_PG` was introduced in #13370 (https://github.com/ray-project/ray/pull/13370) (Jan 2021, @krfricke) as part of placement group support, and that includes documentation in `doc/source/tune/user-guide.rst` (docs about this do not mention anything about the `max_concurrent_trials` arg or its relationship to which searcher you use, though).

**Review by pseudo-rnd-thoughts — APPROVED** [review]
LGTM

**Comment by saschwartz** [comment]
Hi @pseudo-rnd-thoughts , thanks for the review! Just wanted to check what I might need to do to merge this. I reviewed https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst which makes it sounds like Ray maintainers have the ball, but let me know if there's steps outside of that doc I need to take, cheers.
